### PR TITLE
Fix query block duplicate group block rename controls for WP 6.4

### LINKED ISSUE #55517: "Rename" control for blocks within the Query Loop block are repeated
### Description

The rename option is getting repeated in query loops.

CC @andrewserong 

### Step-by-step reproduction instructions

1. Use the query loop and create a stack within
2. Open the options on the stack

### Screenshots, screen recording, code snippet

image
<details><summary>Block Markup</summary>
<p>

```
<!-- wp:query {"queryId":0,"query":{"perPage":10,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false},"enhancedPagination":true,"layout":{"type":"constrained"}} -->
<div class="wp-block-query"><!-- wp:post-template -->
<!-- wp:group {"style":{"spacing":{"blockGap":"var:preset|spacing|default"}},"layout":{"type":"flex","orientation":"vertical","flexWrap":"wrap","justifyContent":"stretch"}} -->
<div class="wp-block-group"><!-- wp:group {"style":{"spacing":{"blockGap":"var:preset|spacing|tight"}},"layout":{"type":"flex","orientation":"vertical"}} -->
<div class="wp-block-group"><!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group"><!-- wp:post-date {"style":{"elements":{"link":{"color":{"text":"var:preset|color|brand-dark-default"}}}},"textColor":"brand-dark-default","fontSize":"text-sm"} /-->

<!-- wp:post-author-name {"style":{"elements":{"link":{"color":{"text":"var:preset|color|brand-dark-default"}}}},"textColor":"brand-dark-default","fontSize":"text-sm"} /--></div>
<!-- /wp:group -->

<!-- wp:post-title /--></div>
<!-- /wp:group -->

<!-- wp:post-excerpt {"moreText":"\u003cem\u003eRead more\u003c/em\u003e","showMoreOnNewLine":false,"style":{"elements":{"link":{"color":{"text":"var:preset|color|dark-lighter"}}}}} /-->

<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|close","bottom":"var:preset|spacing|close","left":"var:preset|spacing|close","right":"var:preset|spacing|close"}}},"backgroundColor":"dark-contrast","layout":{"type":"flex","flexWrap":"nowrap"},"fontSize":"text-sm"} -->
<div class="wp-block-group has-dark-contrast-background-color has-background has-text-sm-font-size" style="padding-top:var(--wp--preset--spacing--close);padding-right:var(--wp--preset--spacing--close);padding-bottom:var(--wp--preset--spacing--close);padding-left:var(--wp--preset--spacing--close)"><!-- wp:paragraph -->
<p>Tags:</p>
<!-- /wp:paragraph -->

<!-- wp:post-terms {"term":"post_tag","style":{"elements":{"link":{"color":{"text":"var:preset|color|brand-dark-default"}}}}} /--></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
<!-- /wp:post-template -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:query-pagination {"layout":{"type":"flex","justifyContent":"space-between"}} -->
<!-- wp:query-pagination-previous /-->

<!-- wp:query-pagination-next /-->
<!-- /wp:query-pagination --></div>
<!-- /wp:group -->

<!-- wp:query-no-results -->
<!-- wp:paragraph {"placeholder":"Add text or blocks that will display when a query returns no results."} -->
<p></p>
<!-- /wp:paragraph -->
<!-- /wp:query-no-results --></div>
<!-- /wp:query -->
```
</p>
</details> 

### Environment info

- WP 6.3.2
- GB 16.8.1
- Custom theme

### Please confirm that you have searched existing issues in the repo.

No

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

No

### PR DESCRIPTION
**Pull request by getdave**
## What?

Fixes the repeated `Rename` menu item in the block options menu when the Group block is within a Query Loop block. 

Fixes https://github.com/WordPress/gutenberg/issues/55517

## Why?

Because you only want to rename once, rename once, rename once, rename once...

## How?

Removes the reliance on `Block.Edit` hook and instead moves the menu item directly into the `BlockSettingsMenuControlsSlot`.

Why is this ok?

For 6.4 this feature exists only for the Group block. However in Gutenberg `trunk` [the feature is enabled for (nearly) all blocks](https://github.com/WordPress/gutenberg/pull/54426).

This means the feature is an integral part of the Editor and so using a hook no longer makes sense. Indeed, we agreed on such an approach in https://github.com/WordPress/gutenberg/pull/55250#issuecomment-1759928887.

## Testing Instructions

- Create ~lots of~ 31 Posts (I used Fakerpress)
- Add a Query loop block
- Add a Group block within the Post Template
- Add another Group block _outside_ of the Query loop.
- Click on the Group inside the Query Loop and trigger the block options menu in the list view. You should see only a _single_ `Rename` option.
- Do the same on the Group outside the loop.
- Validate that the renaming functionality otherwise behaves as per the `wp/6.4` branch.
- Also check that `Rename` is only available for Group
- Check that you can still rename the block via the inspector controls `Advanced -> Rename`.

### CODE REVIEW

File: packages/block-editor/src/components/block-rename/empty-string.js [hunk 1]
```diff
@@ -0,0 +1 @@
+export const emptyString = ( testString ) => testString?.trim()?.length === 0;
```
Comment (scruffian) [inline packages/block-editor/src/components/block-rename/empty-string.js]: Should this be called `isEmptyString`?

**Review by scruffian — APPROVED** [review]
One nit, this looks good to me.

File: packages/block-editor/src/components/block-settings-menu-controls/index.js [hunk 2]
```diff
@@ -44,7 +46,13 @@ const BlockSettingsMenuControlsSlot = ( {
 	);
 
 	const { canLock } = useBlockLock( selectedClientIds[ 0 ] );
+	const { canRename } = useBlockRename( selectedBlocks[ 0 ] );
 	const showLockButton = selectedClientIds.length === 1 && canLock;
+	const showRenameButton =
+		selectedClientIds.length === 1 &&
+		// Todo confirm whether following conditional is needed anymore.
+		// clientId === selectedClientIds[ 0 ] &&
+		canRename;
```
Comment (mikachan) [inline packages/block-editor/src/components/block-settings-menu-controls/index.js]: Just checking if this `todo` was for future reference, or to figure out before 6.4 is released?

**Review by mikachan — COMMENTED** [review]
Thank you for this! Also looks good to me, I've left one inline question.

**Review by mikachan — COMMENTED** [review]
Just noticed that [this test failure](https://github.com/WordPress/gutenberg/actions/runs/6639154329/job/18036950163?pr=55604#step:7:155) may be a genuine failure related to these changes.

File: packages/block-editor/src/components/block-settings-menu-controls/index.js (hunk 2 above)
Comment (getdave) [reply packages/block-editor/src/components/block-settings-menu-controls/index.js]: Good catch. I believe we can remove it as now we've moved the code location there's no need to check the menu applies to the current block.

**Comment by getdave** [comment]
> Just noticed that [this test failure](https://github.com/WordPress/gutenberg/actions/runs/6639154329/job/18036950163?pr=55604#step:7:155) may be a genuine failure related to these changes.

And this is why I wrote good tests for this feature 🎉 

Yes we need to reinstate the inspector control to allow renaming. Probably requires re-adding the hook.

File: packages/block-editor/src/components/block-rename/empty-string.js (hunk 1 above)
Comment (getdave) [reply packages/block-editor/src/components/block-rename/empty-string.js]: ✅

**Comment by getdave** [comment]
Note that when this PR merges we will want to create a PR which merges the changes here with those in `trunk` to enable the rename for _all_ blocks.

This could be a relatively complex merge operation due to how `trunk` has diverged from 6.4 branch on this feature.

**Comment by t-hamano** [comment]
I had made #55629, almost similar to this PR, to solve a different problem 😄 This PR should also resolve the PR reported in #55612.

**Review by mikachan — APPROVED** [review]
Changes look good to me and tests are happy! 🚢

**Review by scruffian — APPROVED** [review]
LGTM

**Comment by SiobhyB** [comment]
I've gone ahead to remove the `Backport to WP Beta/RC` label as this has been merged directly to `wp/6.4`.

**Pull request merged**